dan_layer/engine/src/runtime/tracker.rs

```diff
@@ -502,16 +502,38 @@ impl<TTemplateProvider: TemplateProvider<Template = LoadedTemplate>> StateTracke
     ) -> Result<FinalizeTracker, RuntimeError> {
         // Resolve the transfers to the fee pool resource and vault refunds
         let finalized_fee = self.finalize_fees(&mut substates_to_persist)?;
+
+        let transaction_receipt = substates_to_persist
+            .entry(SubstateAddress::TransactionReceipt(self.transaction_hash().into()))
+            .and_modify(|x| {
+                x.into_transaction_receipt().and_then(|x| {
+                    Some(TransactionReceipt {
+                        transaction_hash: x.transaction_hash,
+                        events: x.events.clone(),
+                        logs: x.logs.clone(),
+                        fee_receipt: Some(finalized_fee.clone()),
+                    })
+                });
+            })
+            .or_insert_with(|| panic!("Failed to get transaction receipt"));
```

Comment:
This isnt doing anything. `and_modify` passes in a mutable reference `FnOnce(&mut V)`which you should mutate. This is mapping reference to a new TransactionReceipt and then discarding it (semi-colon at the end of the closure).

I suggest something like this
```suggestion
        let transaction_receipt_mut = substates_to_persist
            .get_mut(&SubstateAddress::TransactionReceipt(self.transaction_hash().into()))
            .and_then(|s| s.as_transaction_receipt_mut())
            .expect("invariant violation: expected transaction receipt to be added to substates to persist but it wasnt");
            
        transaction_receipt_mut.fee_receipt = Some(finalize_fee.clone());
            let events = transaction_receipt_mut
            .events
            .clone();
        let logs = transaction_receipt_mut
            .logs
            .clone();
```

Reply:
In `impl SubstateValue`
```rust
pub fn as_transaction_receipt_mut(&mut self) -> Option<&mut TransactionReceipt> {
 /// ...
}